feat(gui): add folder picker (fixes #2958)

mathias4833 opened the pull request:
### Purpose

This change introduces a folder picker to the GUI, as suggested in issue #2958. It aims to make it easier for users (especially less technical ones) to select deeply nested folders without having to type them manually.

The goal is to make folder setup more intuitive and beginner-friendly.

It uses the existing `/rest/system/browse` endpoint to retrieve folders. There are no changes on the backend, everything happens on the frontend.

### Testing

- Tested on both Light, Dark and Black themes
- Tested on Linux and Windows
- Tested on Firefox and Chrome

### Screenshots

![File Picker Light Theme](https://github.com/user-attachments/assets/ab96525a-ad54-4286-9599-539d5ea111b5)

![File Picker Dark Theme](https://github.com/user-attachments/assets/cf9a7033-688f-4ccd-a5db-1beb0a7a5b85)

![File Picker Black Theme](https://github.com/user-attachments/assets/2a5d4262-72d8-4eb2-b0e1-4330d8050a37)

mathias4833 committed:
feat(gui): add folder picker (fixes #2958)

tomasz1986 commented:
This is a nice feature to have, especially for the non-tech-savvy user 🙂.

Just a few comments after a very quick test (on Windows):

1. If the initial folder path is empty (which is the default when creating new folders) and you click "Browse", the path in the pop-up window results in `\` and doesn't allow to actually browse any folders.

2. Would it be possible to strip the trailing slash after selecting a folder (so that the slash doesn't end up in the final folder path)?

3. The "Show hidden files" checkbox is confusing. It's not about hidden files but dotfiles. On Windows, files/folders that begin with `.` aren't anything special. On the other hand, the actual hidden files and folders are always visible in the rendered tree. Honestly, I'd suggest not to implement this functionality, as the concept of what a hidden file is differs depending on the OS (see https://github.com/syncthing/syncthing/pull/6535 for a similar case that was eventually dropped).

4. The visuals of the tree could probably be improved a bit, as currently you've got a bordered box rendered inside yet another bordered box. I think it could probably look similar to the Restore Versions tree. This is minor though and could easily be improved later.

tomasz1986 commented on gui/default/syncthing/folder/editFolderModalView.html:
```diff
@@ -30,10 +30,18 @@
+              </datalist>
+              <span class="input-group-btn">
+                <button class="btn btn-default" type="button" ng-click="openFolderPicker()" ng-disabled="editingFolderExisting()">
+                  <i class="fas fa-folder-open"></i>
```
```suggestion
                  <span class="fa fa-folder-open"></span>
```

Please change all `<i>` tags used for icons to `<span>` and also use just `fa` instead of `fas` (as the latter syntax with `fas` is outdated).

mathias4833 committed:
feat(gui): address folder picker review comments

mathias4833 committed:
feat(gui): display folder structure as a tree

mathias4833 commented:
Thank you for your feedback! I've made a few adjustments based on your comments.

Regarding the design, I hadn't had the chance to check the Restore Versions modal yet, so I wasn't aware we were already using `fancytree`. I agree that using it will help make the interface more polished, so I've integrated it to build the folder structure.

Currently, when a folder doesn't exist, it is "added" to the tree to allow the user to fill in an arbitrary path in the input field. However, we can't create a folder by clicking through the existing directories in the tree. I haven't found a simple way to implement this behavior yet. Is the current approach sufficient ? 

Let me know if any further adjustments are needed!

![20250406_13h53m11s_grim](https://github.com/user-attachments/assets/f05de4c2-ae66-4723-a844-8111a881cf65)
![20250406_13h51m32s_grim](https://github.com/user-attachments/assets/234c7a71-384d-41a1-be7b-a087c908de00)

tomasz1986 commented on gui/default/syncthing/folder/folderPickerModalView.html:
```diff
@@ -0,0 +1,30 @@
+    </button>
+    <button class="btn btn-primary btn-sm" id="folderPickerSelect" type="button">
+      <span class="fa fa-check"></span>&nbsp;<span translate>Select</span>
+    </button>
```
Please change the button order, so that "Select" comes before "Cancel".

tomasz1986 commented on gui/default/syncthing/folder/folderPickerModalView.html:
```diff
@@ -0,0 +1,30 @@
+            <span translate>Loading data...</span>
+          </td>
+        </tr>
+        </tbody>
```
Please fix the indentation (`<tbody>` and `<tr>` are on the same level right now).

tomasz1986 commented on gui/default/assets/css/overrides.css:
```diff
@@ -401,6 +372,15 @@ ul.three-columns li, ul.two-columns li {
+/* Maintain consistent padding with the restoreTree component */
+#folderTree td {
+    padding: 2px 0;
+}
```
Maybe move all tree-related code to `tree.css`? This should help with future maintainability.

tomasz1986 commented:
I have tested the current version. It seems to be working as expected, and the tree looks very nice visually 🙂.

1. I haven't tried it, but what is going to happen when you administer Syncthing remotely? Is this going to pull the folder structure from the remote device, or the local device?

2. Very minor, but you seem to be creating files with no new line at the end. The current convention is to have a new line there, so if possible, please add it to all the newly added files that currently end without it.

tomasz1986 commented on gui/default/assets/css/overrides.css:
```diff
@@ -553,3 +533,9 @@ html[lang|="ko"] i {
+
+/* Folder Picker */
+.directory-list {
+    max-height: 400px;
```
Is this value arbitrary? If yes, then I think that it could probably be possible to calculate actual pixels here. This is what I did when modernising the Restore Versions modal, i.e. I just used the browser's developer tools to inspect and calculate the maximum possible height.

mathias4833 replied:
Oops, this shouldn't have been here. Thanks for spotting it. I was actually using the values you calculated for the restore version component (using a shared class). It worked because the two modals are exactly the same height: the tree component, a divider and an input field.

Looking back, using a shared class for setting `max-height` could cause issues if the layout changes, so I've moved to using specific ids for each modal.

mathias4833 committed:
fix(gui): improve code as per peer review

- Changed button order: "Select" now appears before "Cancel"
- Fixed HTML indentation
- Moved tree-related styles to `tree.css`
- Refactored tree.css to improve style sharing and reuse
- Added missing newline at the end of new files for consistency

mathias4833 commented:
> I haven't tried it, but what is going to happen when you administer Syncthing remotely? Is this going to pull the folder structure from the remote device, or the local device?

When administering Syncthing remotely, the folder structure will be pulled from the device where syncthing is running. The current implementation still uses the [`rest/system/browse`](https://docs.syncthing.net/rest/system-browse-get) endpoint. The code leverages the `lazyLoad` method from FancyTree to dynamically load folders, so a request is triggered each time a folder is opened for the first time.

tomasz1986 commented on gui/default/assets/css/tree.css:
```diff
@@ -56,3 +56,91 @@
     word-break: break-all;
 }
+
+.fancytree-temp .fancytree-title {
```
```suggestion
.folderTree-new-folder {
```

You can apply this to `tr` directly.

tomasz1986 commented on gui/default/syncthing/folder/folderPickerModalController.js:
```diff
@@ -0,0 +1,171 @@
+                nextNode = currentNode.addChildren({
+                    title: key,
+                    key: joinPath(currentNode.key, key),
+                    extraClasses: "fancytree-temp",
```
```suggestion
                    extraClasses: "folderTree-new-folder",
```

tomasz1986 commented on gui/default/assets/css/tree.css:
```diff
@@ -56,3 +56,91 @@
 }
+
+.fancytree-temp .fancytree-title {
+    opacity: 0.6;
```
```suggestion
    color: #626262
```

The opacity makes the contrast ratio fail WCAG AA, so I would suggest to use these colours instead. The value above is for the light theme. The values for the black and dark themes are `color: #aaa`. All these are already used and come from `help-block`.

tomasz1986 commented on gui/default/assets/css/tree.css:
```diff
@@ -56,3 +56,91 @@
+
+.fancytree-temp .fancytree-title {
+    opacity: 0.6;
+    font-style: italic;
```
Italic is problematic for some languages (see https://github.com/syncthing/syncthing/blob/5f5d672a7d16c4e38a84e8af2feb7c04e1287212/gui/default/assets/css/overrides.css#L545-L550).

Maybe add

```
html[lang|="zh"] .folderTree-new-folder,
html[lang="ja"] .folderTree-new-folder,
html[lang|="ko"] .folderTree-new-folder {
    font-style: normal;
}
```

? The other way is to use `:not`, but the override seems simpler…

tomasz1986 commented on gui/default/syncthing/folder/folderPickerModalController.js:
```diff
@@ -0,0 +1,171 @@
+                    key: joinPath(currentNode.key, key),
+                    extraClasses: "fancytree-temp",
+                    folder: true,
+                    tooltip: $translate.instant('Folder will be created')
```
Just wondering if it would be possible to use a Bootstrap tooltip here. Normally, it is done by attributes in the HTML, e.g.

```
<span tooltip data-original-title="{{'Folder will be created' | translate}}">
```

mathias4833 committed:
fix(gui): Update new folder style and height calc

mathias4833 committed:
feat(gui): use Bootstrap tooltip

mathias4833 committed:
chore(gui): add missing semicolons for consistency

tomasz1986 commented on gui/default/syncthing/folder/folderPickerModalView.html:
```diff
@@ -0,0 +1,30 @@
+<modal id="folderPicker" status="default" heading="{{'Select Folder' | translate}}" icon="fa fa-folder-open" large="yes"
+       closeable="yes">
```
Thanks for all the fixes. This looks great and works as expected for me 🙂.

Just one last suggestion, maybe put all this on the same line? All other modal HTML files have it like that, regardless of the length.

```suggestion
       <modal id="folderPicker" status="default" heading="{{'Select Folder' | translate}}" icon="fa fa-folder-open" large="yes" closeable="yes">
```

mathias4833 committed:
fix(gui): fix indentation for consistency

tomasz1986 commented:
Can you update the screenshots to match the final version? If possible, using both the light and dark themes.

acolomb reviewed (commented):
Really nice, wow!

Are you aware that on Windows, both `/` *and* `\` serve as path separators?  The one stored in the GUI variable is only the *default* for the operating system Syncthing is running on.  But either style is acceptable on Windows, and they may even be mixed in a single path.  How does your tree code handle that, will it recognize those paths as well?  Kind of an edge case, but I do encounter it regularly.

I had no chance to test this on Windows, so couldn't check for myself.  But on Linux it works fine, just one comment regarding the tilde handling.

Otherwise, I've only glanced over the code.  Will need some more time to grok it fully, but OTOH I don't mind much as long as it works.  And what I've seen code-wise looks clean and thorough.  So together with @tomasz1986's review, I'm okay with this in general.

mathias4833 committed:
fix(gui): handle path edge cases

- On Windows, '/' is replaced with '\' to handle paths correctly.
- Fixed path edge cases such as '.' (current directory), '..' (parent directory), and consecutive slashes (e.g. '///').
- Tilde expansion now only applies when '~' is the first component of the path.

mathias4833 commented:
I wasn't aware that `/` could be used on Windows too, thanks for pointing that out!

I had indeed missed  a few edge cases. The latest commit should take care of that. Paths with `.` and `..`, as well as those with multiple consecutive separators (like `~/Folder1////.././`), should now be handled correctly.

rasa commented:
> I wasn't aware that `/` could be used on Windows too, thanks for pointing that out!

Just to be clear, @acolomb is referring to passing filenames to a Win32 API. Slashes are still rejected by Windows Explorer, so we may want to *show* backslashes to the user to not confuse them, but we are free to use only slashes internally.

tomasz1986 commented:
> Slashes are still rejected by Windows Explorer, so we may want to _show_ backslashes to the user to not confuse them, but we are free to use only slashes internally.

Forward slashes actually do work in Windows Explorer (if input manually), at least in Windows 10 and above, but the default is still backslash everywhere, so people using Windows are just accustomed to seeing backslash as the path separator.

rasa committed:
Merge branch 'main' into feat/folder-picker

calmh committed:
Merge branch 'main' into feat/folder-picker

calmh commented:
So, anyone who's done the review up for an approval?

tomasz1986 commented:
I can approve, but what about the failing checks?

calmh commented:
Unrelated, will work on retry... the DCO would be nice @mathias4833 but given the age here I think we can consider it grandfathered in

mathias4833 commented:
Both options are fine with me. I can add the DCO sign-offs if needed but I'm a bit unsure about the cleanest way to do that since there have been a few merges from main into this branch. If you'd prefer that, I'd appreciate some guidance on the best way to go about it: the DCO check suggests `git rebase HEAD~11 --signoff`, but is that still the right approach here given the merges from main since then ?

calmh committed:
Merge branch 'main' into feat/folder-picker

calmh commented:
You might get a few merge conflicts given the history here, I'd probably do a squash merge into a new branch instead, but honestly, if it's tricky I'd just leave it out at this point and we add it in the merge.

mathias4833 commented:
Makes sense, thanks. I’ll leave it as-is and we can add the DCO at merge time if that works

acolomb commented on gui/default/syncthing/folder/folderPickerModalController.js:
```diff
@@ -0,0 +1,191 @@
+        }
+
+        function splitAndNormalize(path) {
+            if ($scope.$parent.version.os.toLowerCase() === "windows") {
+                // Since both '/' and '\' are valid path separators on Windows
```
Could simply check if the default OS path separator is a backslash, then replace all forward slashes with it.  We know that only Windows uses backslash as a default, and no other OSes do.  So the assumption "if the default is backslash, normalize forward slashes to it" holds.  And we have a common source of truth, the path separator info from the API.

rasa replied:
Can't we just use forward slashes? Windows allows forward slashes on API calls. There is no reason to use backslashes, unless we think that will confuse Windows users.

acolomb replied:
I do see some potential for confusion, so as long as it doesn't cause lots of trouble, it's alright to normalize to backslashes on Windows. We probably get the paths from the API endpoint in "native" form as well. Can't find it right now, but I think there was a reference above to an older issue where the path separator was added to the API response, and that seemed reasonable to me.

acolomb commented on gui/default/syncthing/folder/folderPickerModalController.js:
```diff
@@ -0,0 +1,191 @@
+                params: {current: addTrailingSeparator(path)}
+            });
+            return res.data.map(dir => ({
+                title: formatDirectoryName(dir),
```
Isn't this a bit wasteful, going over each entry again, completely splitting all components just to retrieve the final, individual part?  At the least, the result from `splitAndNormalize()` could be re-used, pulling out the final component after normalization.  (For the title alone, a reverse search for the path separator would be sufficient though.)

acolomb commented on gui/default/syncthing/folder/folderPickerModalController.js:
```diff
@@ -0,0 +1,191 @@
+        }
+
+        function formatDirectoryName(path) {
+            return splitAndNormalize(path).pop() || path;
```
What is the fallback supposed to help with?  Can you describe the circumstances where pop() doesn't return a valid string?

acolomb commented on gui/default/syncthing/folder/folderPickerModalController.js:
```diff
@@ -0,0 +1,191 @@
+                    indentation: 24,
+                },
+                strings: {
+                    loading: $translate.instant("Loading data..."),
+                    loadError: $translate.instant("Failed to load data"),
```
Please run `go run build.go translate` to get the newly added strings into the template along with this PR.

acolomb replied:
Just pushed an update for this.

acolomb reviewed (commented):
Finally found some time to review this again, sorry for the long wait.  I'm very happy with it overall, good interface and a clean implementation.  Some nits noted in inline comments, for further consideration.

Observations:
1. When a path is entered in the text field, it is correctly highlighted in the tree.  When clicking in the tree and then returning focus back to the text entry, the highlighting is lost though.  Would that be hard to fix?

2. There is no background color defined in the default theme for `.fancytree-focused`.  It's present in all three themes, but when a custom theme is added, there is no default style (in `tree.css`).  I know this fault was already present before your PR, but could you please fix that while already touching the tree styles?

3. One can exit / cancel the modal dialog with the Escape key, but after that, the same key does not work to close the folder adding modal anymore.  Can this be fixed, or is it a general shortcoming of our Angular version?

Further ideas:
* I think it would be sensible to initialize the folder picker at the "home" folder (`system.tilde`).  Sure, the regular path entry starts out empty, which the picker interprets as the root path.  But especially on UNIXy systems, the root path is seldomly the right place to look for Syncthing folders.  Other environments (think custom NAS OSes) usually set a sensible `$HOME`, but finding that from the root path might be unexpectedly hard for users.  And entering a tilde to get there is not something most users will do as first reflex.  This could easily be changed in this PR.

* For a follow-up if you like:  With the regular path entry, we switched from replacing the tilde upon entry to keeping it in the config and replacing on configuration load, some time ago.  I think the folder picker should do the same, allowing the tilde to stay in place.  This is especially important for portable installations for example, where `$HOME` will be dynamically defined before starting Syncthing, and the config doesn't need adaptation to find the paths again.

  So the idea is to allow three possible "root" paths:  If the entry starts with `/` (or a drive letter), current behavior is fine.  If it starts with a tilde `~/`, you take that as the root element of the tree and only list paths below `$HOME`.  If it starts with `./`, paths are treated as relative to Syncthing's working directory, just as they would apply when putting relative paths in the config file.  Finding the working directory might need an API addition, I suppose.

acolomb committed:
Regenerate translation strings.

Signed-off-by: André Colomb <src@andre.colomb.de>

acolomb committed:
Move hover / focus CSS overrides from light to default theme.

For custom themes, no special highlight background color is defined
for hovering or an explicitly selected (focused) FancyTree row.  By
moving it to the tree.css overrides, the setting applies by
default (including the light theme), unless specifically overridden by
another theme.

Signed-off-by: André Colomb <src@andre.colomb.de>
